File: adk/go/weil_go/collections/map.go
package collections

import (
	"encoding/json"
	"fmt"

	"github.com/weilliptic-public/wadk/adk/go/weil_go/runtime"
)

// WeilMap is a key-value map collection that can be persisted to the contract's state tree.
// It supports generic key types (must be comparable) and value types.
type WeilMap[K comparable, V any] struct {
	StateId WeilId `json:"state_id"`
}

// NewWeilMap creates a new WeilMap with the given state identifier.
func NewWeilMap[K comparable, V any](stateId WeilId) *WeilMap[K, V] {
	return &WeilMap[K, V]{
		StateId: stateId,
	}
}

// BaseStatePath returns the base path for this map in the state tree.
func (m *WeilMap[K, V]) BaseStatePath() string {
	return m.StateId.String()
}

// StateTreeKey generates a state tree key for the given key suffix.
func (m *WeilMap[K, V]) StateTreeKey(suffix *K) string {
	serializedSuffix, _ := json.Marshal(suffix)
	return fmt.Sprintf("%s_%s", m.BaseStatePath(), string(serializedSuffix))
}

// Insert adds or updates a key-value pair in the map.
// If the key already exists, the value will be overwritten.
func (m *WeilMap[K, V]) Insert(key *K, val *V) {
	runtime.WriteCollectionEntry[V]([]byte(m.StateTreeKey(key)), val)
}

// Get retrieves the value associated with the given key.
// Returns an error if the key is not found in the map.
func (m *WeilMap[K, V]) Get(key *K) (*V, error) {
	return runtime.ReadCollection[V]([]byte(m.StateTreeKey(key)))
}

// Remove removes the key-value pair associated with the given key from the map.
// Returns the removed value if it existed, or an error if the key was not found.
func (m *WeilMap[K, V]) Remove(key *K) (*V, error) {
	return runtime.DeleteCollectionEntry[V]([]byte(m.StateTreeKey(key)))
}

File: adk/asc/weil-sdk/assembly/collections/map.ts
import { WeilId } from './weil-id'
import { Memory } from '../runtime'
import { JSON } from 'json-as'

/**
 * A persistent key-value map collection that stores data in the Weil runtime state.
 * 
 * @template K - The type of keys in the map
 * @template V - The type of values in the map
 */
@json
export class WeilMap<K, V> {
  /** The unique identifier for this map's state storage */
  stateId: WeilId

  /**
   * Creates a new WeilMap instance with the specified state ID.
   * 
   * @param stateId - The unique identifier for this map's state storage
   */
  constructor(stateId: WeilId) {
    this.stateId = stateId
  }

  /**
   * Gets the base state path for this map.
   * 
   * @returns The string representation of the state ID
   */
  private getBaseStatePath(): string {
    return this.stateId.toString()
  }

  /**
   * Generates a unique state tree key for a given key suffix.
   * 
   * @param suffix - The key to generate a state tree key for
   * @returns A unique string key for state storage
   */
  private getStateTreeKey(suffix: K): string {
    return [this.getBaseStatePath(), JSON.stringify(suffix)].join('_')
  }

  /**
   * Retrieves the value associated with the given key.
   * 
   * @param key - The key to look up
   * @returns The value associated with the key, or null if not found
   */
  get(key: K): V | null {
    return Memory.readCollection<V>(this.getStateTreeKey(key))
  }

  /**
   * Checks whether a key exists in the map.
   * 
   * @param key - The key to check
   * @returns True if the key exists, false otherwise
   */
  has(key: K): boolean {
    return this.get(key) !== null
  }

  /**
   * Sets a key-value pair in the map.
   * 
   * @param key - The key to set
   * @param value - The value to associate with the key
   */
  set(key: K, value: V): void {
    return Memory.writeCollection<V>(this.getStateTreeKey(key), value)
  }

  /**
   * Deletes a key-value pair from the map.
   * 
   * @param key - The key to delete
   * @returns The value that was associated with the key, or null if not found
   */
  delete(key: K): V | null {
    return Memory.deleteCollection<V>(this.getStateTreeKey(key))
  }
}
